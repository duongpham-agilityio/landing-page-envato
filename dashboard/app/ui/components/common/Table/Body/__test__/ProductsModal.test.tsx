import { render, screen, fireEvent, waitFor } from '@testing-library/react';
import { ProductModal } from '..';
import userEvent from '@testing-library/user-event';

describe('Product Modal', () => {
  const mockProduct = {
    id: '123',
    name: 'Tivi',
    price: 1111,
    stock: 11,
    imageURLs: ['image.png'],
  };

  it('match to snapshot', () => {
    const { container } = render(
      <ProductModal isDelete product={mockProduct} />,
    );

    expect(container).toMatchSnapshot();
  });

  it('should render delete confirmation correctly', async () => {
    render(<ProductModal isDelete product={mockProduct} />);

    // Ensure delete confirmation message is rendered
    expect(
      screen.getByText(/Are you sure delete the product with id/i),
    ).toBeInTheDocument();
  });

  it('should call onDeleteProduct on delete button click', async () => {
    const mockOnDeleteProduct = jest.fn();
    render(
      <ProductModal
        isDelete
        product={mockProduct}
        onDeleteProduct={mockOnDeleteProduct}
      />,
    );

    // Click the delete button
    fireEvent.click(screen.getByText('Delete'));

    // Ensure onDeleteProduct is called
    expect(mockOnDeleteProduct).toHaveBeenCalled();
  });

  it('should call onCloseModal on cancel button click', async () => {
    const mockOnCloseModal = jest.fn();
    render(
      <ProductModal
        isDelete
        product={mockProduct}
        onCloseModal={mockOnCloseModal}
      />,
    );

    // Click the cancel button
    fireEvent.click(screen.getByText('Cancel'));

    waitFor(() => {
      // Ensure onCloseModal is called
      expect(mockOnCloseModal).toHaveBeenCalled();
    });
  });

  it('handleSubmitForm should update product, reset form, and close modal', async () => {
    const onUpdateProductMock = jest.fn();
    const onCloseModalMock = jest.fn();

    render(
      <ProductModal
        product={mockProduct}
        isDelete={false}
        onUpdateProduct={onUpdateProductMock}
        onCloseModal={onCloseModalMock}
      />,
    );

    const inputField = screen.getByTestId('edit-field-name');

    await fireEvent.change(inputField, { target: { value: 'NewValue' } });

    await userEvent.click(screen.getByTestId('submit-product-form'));

    // Assertions
    waitFor(() => {
      expect(onUpdateProductMock).toHaveBeenCalled();
      expect(onCloseModalMock).toHaveBeenCalled();
    });
  });
});
