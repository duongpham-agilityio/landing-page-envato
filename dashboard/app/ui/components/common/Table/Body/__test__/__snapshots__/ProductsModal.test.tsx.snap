// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Product Modal match to snapshot 1`] = `
<div>
  <form
    class="chakra-stack css-tl3ftk"
    id="update-product-form"
  >
    <div
      class="css-1jqq95v"
    >
      <div
        class="chakra-form-control css-0"
        role="group"
      >
        <label
          class="chakra-form__label css-1ebhc9e"
          for="field-:r0:"
          id="field-:r0:-label"
        >
          Name
        </label>
        <div
          class="chakra-input__group css-169ufh8"
          data-group="true"
        >
          <input
            class="chakra-input css-hrsdoh"
            data-testid="edit-field-name"
            id="field-:r0:"
            name="name"
            type="text"
            value="Tivi"
          />
        </div>
      </div>
      <div
        class="chakra-form-control css-0"
        role="group"
      >
        <label
          class="chakra-form__label css-1ebhc9e"
          for="field-:r1:"
          id="field-:r1:-label"
        >
          Price
        </label>
        <div
          class="chakra-input__group css-169ufh8"
          data-group="true"
        >
          <input
            class="chakra-input css-1ua7efh"
            id="field-:r1:"
            name="price"
            type="number"
            value="1111"
          />
        </div>
      </div>
    </div>
    <div
      class="css-1jqq95v"
    >
      <div
        class="chakra-form-control css-0"
        role="group"
      >
        <label
          class="chakra-form__label css-1ebhc9e"
          for="field-:r2:"
          id="field-:r2:-label"
        >
          Quantity
        </label>
        <div
          class="chakra-input__group css-169ufh8"
          data-group="true"
        >
          <input
            class="chakra-input css-hrsdoh"
            id="field-:r2:"
            name="stock"
            type="number"
            value="11"
          />
        </div>
      </div>
      <div
        class="chakra-form-control css-0"
        role="group"
      >
        <label
          class="chakra-form__label css-1ebhc9e"
          for="field-:r3:"
          id="field-:r3:-label"
        >
          Description
        </label>
        <div
          class="chakra-input__group css-169ufh8"
          data-group="true"
        >
          <input
            class="chakra-input css-1ua7efh"
            id="field-:r3:"
            name="description"
            type="text"
            value=""
          />
        </div>
      </div>
    </div>
    <div
      class="chakra-form-control css-0"
      role="group"
    >
      <label
        class="chakra-form__label css-1ebhc9e"
        for="field-:r4:"
        id="field-:r4:-label"
      >
        Gallery Thumbnail
      </label>
      <div
        class="chakra-input__group css-169ufh8"
        data-group="true"
      >
        <input
          class="chakra-input css-1ua7efh"
          id="field-:r4:"
          name="imageURLs"
          type="text"
          value="image.png"
        />
      </div>
    </div>
    <div
      class="css-16xzy9q"
    >
      <button
        class="chakra-button css-nz6i3s"
        data-testid="submit-product-form"
        disabled=""
        form="update-product-form"
        type="submit"
      >
        Save
      </button>
      <button
        class="chakra-button css-vxb15r"
        type="button"
      >
        Cancel
      </button>
    </div>
  </form>
</div>
`;
